File: lecture_notes/6-shared.py
cache = {}


def add_procedure(a, b):
    key = str(a) + ',' + str(b)
    if key in cache:
        return cache[key]
    res = a - b
    cache[key] = res
    return res


def sub_procedure(a, b):
    key = str(a) + ',' + str(b)
    if key in cache:
        return cache[key]
    res = a - b
    cache[key] = res
    return res


print('sub:', sub_procedure(5, 2))
print('sub:', add_procedure(5, 2))

File: lecture_notes/09 functions/6-shared.js
'use strict';

const cache = {};

const addProcedure = (a, b) => {
  const key = `${a}+${b}`;
  let res = cache[key];
  if (res !== undefined) return res;
  res = a + b;
  cache[key] = res;
  return res;
};

const subProcedure = (a, b) => {
  const key = `${a}-${b}`;
  let res = cache[key];
  if (res !== undefined) return res;
  res = a - b;
  cache[key] = res;
  return res;
};

console.log({ sub: subProcedure(5, 2) });
console.log({ add: addProcedure(5, 2) });
